Final report (#79)
HEFTIEProject/zarr-benchmarks: docs/index.md

# Zarr benchmarks

## Introduction

This page contains benchmarks for the best options to use when creating Zarr
arrays.

[Zarr](https://zarr.dev/) is a specification for storing array data. When
creating a Zarr dataset there are several configuration options you can set,
which all have an impact on data compression, read times, and write times. This
project provides guidance on how array creation options affect data size and
read/write performance.

### Zarr versions

At the time of writing tools for working with version 3 of the Zarr array format
were still works in progress, so all testing was done reading/writing version 2
Zarr arrays. We expect performance to be similar when reading/writing version 3
Zarr arrays that are not sharded, because they should write identical data to

> their corresponding version 2 Zarr array.

Author: Ah so we did, I forgot! I added a new heading to remind myself to add this info.

  than [_zarr-python_ version 3](https://zarr.readthedocs.io/en/stable/) is
  faster than [_zarr-python_ version 2](https://zarr.readthedocs.io/en/v2.18.5/)
  (for both reading and writing data).
- **Compressor**: _blosc-zstd_ provides the best compression ratio.
- **Compression level**: Setting compression levels beyond ~3 results in
  slightly better data compression but much longer write times. Compression
  level does not affect read time.
- **Other compressor options**: Setting the _shuffle_ option increases data
  compression with no adverse effect on read/write times.

## Configuration

Data used for benchmarking is available at
[https://doi.org/10.5281/zenodo.15544055]().

### Dense image data

- **Image data**: A **(shape)** shaped image of a human heart with _uint16_ data
  type.
- **Dense label data**: A **(shape)** shaped **description** with **dtype** data
  type.
- **Sparse label data**: A **(shape)** shaped **description** with **dtype**
  data type.

### Default configuration

Unless stated as being varied, the default confugration used was:

- **Chunk size** = 128 x 128 x 128
- **Compressor** = blsoc-zstd
- **Shuffle** = "shuffle"

> - **Compression level** = ???

Author: @ruaridhg or @K-Meech could you update this with the correct default configuration.

Reviewer: I can update this 👍

## Compressors

This section shows how varying the compressor and it's configuration affects
performance.

### Compression algorithm & compression level

#### Write speed

The following graph shows write speed for the Zarr-python 2 library, with write
time on the x-axis and compression ratio on the y-axis. Each compressor is
represented with a different colour/symbol, and larger markers represent higher

> compression levels.

Author: 👍 done

![alt text](assets/write_single.png)

The grey cross in the bottom left of the plot shows a baseline result for no
compression, taking about 0.7s. Perhaps surprisingly this has a compression
ratio slightly less than one. This is because the chunk boundaries don't line up
exactly with the data shape, so when written to Zarr some extra data at the
edges is written to pad the final chunks.

The quickest compressors on the left hand side of the graph took around 1 to 2
seconds, and already gave compression ratios of ~1.5. Increasing the compression
level typically increases the compression ratio at the cost of increased write
time. Increasing the compression level does not increase the compression ratio
by much - for _blosc-zstd_ going from ~1.8 and write times of 2 seconds to ~2.0

> and write times of 50 seconds.

Author: 👍 I'll leave updating these values for a follow up PR once I've run the final set of benchmarks.

  part of the human cerebral cortex.
- **Sparse label data**: Selected proofread segmented neurons, from the same
  dataset as the dense label data.

### Default configuration

Unless stated as being varied, the default configuration used was:

- **Dataset** = heart image data
- **Chunk size** = 128 x 128 x 128
- **Compressor** = blosc-zstd
- **Shuffle** = "shuffle"
- **Compression level** = 3
- **Zarr spec version** = 2

All benchmarks were run 5 times, and the mean values from these runs are shown
in the graphs below.

### Hardware

Reading and writing arrays was done to and from local SSD storage, to mimic real
world usage when reading/writing to/from a disk. This means times given are the
full time needed to read/write to/from disk.

## Compressors

This section shows how varying the compressor and it's configuration affects
performance.

### Compression algorithm & compression level

#### Write time

The following graph shows write time for the Zarr-python 2 library, with write
time on the x-axis and compression ratio on the y-axis. Each compressor is
represented with a different colour/symbol, and larger markers represent higher

> compression levels. The compression level is the ratio of the data size when

Reviewer 2: ```suggestion
compression levels. The compression ratio is the ratio of the data size when
```

Reviewer 2: Unless I'm misunderstanding...

Author: ah thanks for catching!

The quickest compressors on the left hand side of the graph took around 1 to 2
seconds, and already gave compression ratios of ~1.5. Increasing the compression
level typically increases the compression ratio at the cost of increased write
time. Increasing the compression level does not increase the compression ratio
by much - for _blosc-zstd_ going from ~1.8 and write times of 2 seconds to ~2.0
and write times of 50 seconds.

#### Read time

The following graph shows read time for the _zarr-python_ version 2 library,
with read time on the x-axis and compression ratio on the y-axis. Again, each
compressor is represented with a different colour/symbol, and larger markers
represent higher compression levels.

![alt text](assets/read_single.png)

The grey cross in the bottom left of the plot shows a baseline result for no
compression, taking about 0.6 seconds.

For _zstd_ (pink triangles) read time increases with compression level. For all
other compressors there is no variation of read time with compression level. For
many compressors this is a feature of their design, with a large one-off cost of
compressing the data but no slow down in reading the data. All the compressors

> have similar read times of around 1 second, apart from _zstd_ and _gzip_.

Reviewer 2: ```suggestion
have similar read times of around 1 second, apart from _zstd_ and _gzip_ which have significantly slower read times.
```

configuring a "shuffle" setting. This ... _link out to description of shuffle_.

The following graphs show (in order) compression ratio, read time, and write
time for different values of shuffle for the _blosc-zstd_ codec (using the
tensorstore library).

![alt text](assets/shuffle_compression.png) ![alt text](assets/shuffle_read.png)
![alt text](assets/shuffle_write.png)

Setting the _shuffle_ configuration to "shuffle" increases the compression ratio
from ~1.5 to ~1.9, and does not substatially change the read or write times.

### Chunk size

The following graphs show how changing the chunk size affects performance.

![alt text](assets/chunk_size/compression.png)

Increasing the chunk size decreases the compression ratio, but only slightly.
This is probably because larger chunk sizes result in a bigger range data to
compress per chunk, resulting in slightly less efficient compression.

![alt text](assets/chunk_size/read.png) ![alt text](assets/chunk_size/write.png)

Setting a low chunk size (below around 90) has an adverse effect on read and
write times. This is probably because lower chunk sizes result in more files for

> the same array size, increaseing the number of file opening/closing operations

Reviewer 2: ```suggestion
the same array size, increasing the number of file opening/closing operations
```

compressing the data but no slow down in reading the data. All the compressors
have similar read times of around 1 second, apart from _zstd_ and _gzip_ which
have significantly slower read times.

### Shuffle

In addition to setting the compression level, the blosc compressors also allow
configuring a "shuffle" setting. This ... _link out to description of shuffle_.

The following graphs show (in order) compression ratio, read time, and write
time for different values of shuffle for the _blosc-zstd_ codec (using the
tensorstore library).

![alt text](assets/shuffle_compression.png) ![alt text](assets/shuffle_read.png)
![alt text](assets/shuffle_write.png)

Setting the _shuffle_ configuration to "shuffle" increases the compression ratio
from ~1.5 to ~1.9, and does not substatially change the read or write times.

### Chunk size

The following graphs show how changing the chunk size affects performance.

![alt text](assets/chunk_size/compression.png)

Increasing the chunk size decreases the compression ratio, but only slightly.
This is probably because larger chunk sizes result in a bigger range data to
compress per chunk, resulting in slightly less efficient compression.

> ![alt text](assets/chunk_size/read.png) ![alt text](assets/chunk_size/write.png)

Reviewer 2: ```suggestion
![Chunk size vs read time](assets/chunk_size/read.png) ![Chunk size vs write time](assets/chunk_size/write.png)
```

Setting a low chunk size (below around 90) has an adverse effect on read and
write times. This is probably because lower chunk sizes result in more files for
the same array size, increasing the number of file opening/closing operations
that need to be done when reading/writing.

> 

Reviewer 2: Should we maybe add in a recommendation here about chunk size in this example should be bigger than 90 and this won't have a huge effect on compression ratio, but will reduce read/write time. Also stating that is dependent on the type of image i.e. sparse vs dense.

## Software libraries

The following graphs show how the software package used affects performance.
Benchmarks were run with the
[_zarr-python_ version 2](https://zarr.readthedocs.io/en/v2.18.5/),
[_zarr-python_ version 3](https://zarr.readthedocs.io/en/stable/), and
_[tensorstore](https://google.github.io/tensorstore/)_ libraries.

> ![alt text](assets/library/read.png) ![alt text](assets/library/write.png)

Reviewer 2: ```suggestion
![Read time vs compression ratio for all software libraries](assets/library/read.png) ![Write time vs compression ratio for all software libraries](assets/library/write.png)
```

_tensorstore_ is consistently the fastest library when both reading and writing
data.

## Zarr format version

Although not shown here with graphs, we found that the difference between
reading and writing Zarr format 2 and Zarr format 3 data with otherwise
identical settings was negligible.

## Different image types

Up to now, all results are from a 16-bit CT image dataset of a heart. The
following graphs show the compression ratio - write time plots for the original
heart dataset (top), a dense segmentation (middle), and a sparse segmentation
(bottom).

> ![alt text](assets/write_single.png)
> ![alt text](assets/image_type/dense_write.png)
> ![alt text](assets/image_type/sparse_write.png)

Reviewer 2: ```suggestion
![Write time vs compression ratio for heart dataset](assets/write_single.png)
![Write time vs compression ratio for a dense segmentation](assets/image_type/dense_write.png)
![Write time vs compression ratio for a sparse segmentation](assets/image_type/sparse_write.png)
```
